src/main/java/hudson/plugins/git/GitSCM.java

```diff
@@ -1148,6 +1156,30 @@ private void retrieveChanges(Run build, GitClient git, TaskListener listener) th
         }
     }
 
+    private boolean determineRedundantFetch(CloneOption option, RemoteConfig rc) {
+        List<RefSpec> initialFetchRefSpecs = rc.getFetchRefSpecs();
+        boolean isDefaultRefspec = true; // default refspec is any refspec with "refs/heads/" mapping
+        boolean removeSecondFetch = true;
+        if (initialFetchRefSpecs != null) {
+            for (RefSpec ref : initialFetchRefSpecs) {
+                if (!ref.toString().contains("refs/heads")) {
+                    isDefaultRefspec = false; // if refspec is not of default type, preserve second fetch
+                }
+            }
+            if (option == null) {
+                removeSecondFetch = isDefaultRefspec;
+            } else {
+                if (!option.isHonorRefspec()) {
+                    removeSecondFetch = isDefaultRefspec;
+                } else {
+                    removeSecondFetch = true; // avoid second fetch call if honor refspec is enabled
+                }
```

Comment:
This if clause didn't help me to see clearer what you meant. While having a "!" in an `if` clause is perfectly fine, the "!" in the `if-else` might cause confusion. I would have expected 
```
if something is true
   sentence1
else
   sentence2
```
while this piece of code is
```
if something is not false
   sentence2
else
   sentence1
```
My personal preference is to have the first approach in terms of readability. But it's just a matter of personal taste, so take this advice as it is, just an advice.

Reply:
This is a great suggestion @fcojfernandez. I have added this in c2e097d.